src/circuit.cpp

```diff
@@ -1,13 +1,331 @@
+{0xad21b516cbc645ffe34ab5de1c8aef8c_cppui255, 0xd4e7f8d2b51e8e1456adc7563cda206f_cppui255}
+}};
+
+// This is a constant in Ethereum network
+constexpr std::size_t VALIDATORS_MAX_SIZE_LOG2 = 40;
+// constexpr std::size_t VALIDATORS_MAX_SIZE = 1 << VALIDATORS_MAX_SIZE_LOG2;
+// Validator merkleization is an actual sha2-256 hash, so the full tree height is used
+constexpr std::size_t VALIDATORS_TARGET_TREE_HEIGHT = VALIDATORS_MAX_SIZE_LOG2;
+
+constexpr std::size_t BALANCES_PER_LEAF_LOG2 = 2;
+constexpr std::size_t BALANCES_PER_LEAF = 1 << BALANCES_PER_LEAF_LOG2; // // 2 ** BALANCES_PER_LEAF_LOG2
+// Balances are "packed" together (4 balances in one merkle leaf), so their target tree height is smaller by 2
+constexpr std::size_t BALANCES_TARGET_TREE_HEIGHT = VALIDATORS_MAX_SIZE_LOG2 - BALANCES_PER_LEAF_LOG2;
+
+// Theoretically, this should be just VALIDATORS_MAX_SIZE_LOG2; however, this is impractical, as each of these
+// elements are actually represented by an input value (and variable in the circuit assignment table, and everything else)
+// It basically boils down to vector vs. array - SSZ implementations work with vectors with max size bounded by VALIDATORS_MAX_SIZE,
+// but only carrying actual number of validators (~1M by Oct 2023), while circuit has to use array and thus would carry a full
+// VALIDATORS_MAX_SIZE elements - which is a lot. So we're using a smaller "problem size" to carry (and pass aroung) the necessary
+// minimum of "padding" elemetns.
+constexpr std::size_t PROBLEM_SIZE_LOG2 = 4;
+constexpr std::size_t VALIDATORS_COUNT = 1 << PROBLEM_SIZE_LOG2; // 2 ** PROBLEM_SIZE_LOG2
+constexpr std::size_t VALIDATORS_TREE_HEIGHT = PROBLEM_SIZE_LOG2; // 2 ** PROBLEM_SIZE_LOG2
+constexpr std::size_t BALANCES_COUNT = 1 << PROBLEM_SIZE_LOG2; // 2 ** PROBLEM_SIZE_LOG2
+constexpr std::size_t BALANCES_TREE_HEIGHT = PROBLEM_SIZE_LOG2 - BALANCES_PER_LEAF_LOG2;
+constexpr std::size_t BALANCES_LEAFS_COUNT = BALANCES_COUNT / BALANCES_PER_LEAF;
+
+constexpr std::size_t VALIDATOR_FIELDS = 8;
+constexpr std::size_t BEACON_STATE_FIELD_INCLUSION_PROOF_LENGTH = 5;
+constexpr std::size_t VALIDATORS_FIELD_INDEX = 11;
+constexpr std::size_t BALANCES_FIELD_INDEX = 12;
+
+constexpr std::size_t BEACON_BLOCK_FIELDS_COUNT = 5;
+
+constexpr bool BYTE_ORDER_MSB = true;
+constexpr bool BYTE_ORDER_LSB = false;
+
+bool is_same(block_type block0, block_type block1){
+    return block0[0] == block1[0] && block0[1] == block1[1];
+}
+
+template <std::size_t LayerSize>
+block_type hash_layer(std::array<block_type, LayerSize> input, size_t layer){
+    constexpr size_t NextLayerSize = (LayerSize % 2 == 0) ? LayerSize / 2 : (LayerSize / 2 + 1);
+    std::array<block_type, NextLayerSize> next_layer;
+
+    for (std::size_t leaf_index = 0; leaf_index < LayerSize / 2; leaf_index++) {
+        next_layer[leaf_index] = hash<hash_type>(input[2 * leaf_index], input[2 * leaf_index + 1]);
+    }
+    if (LayerSize % 2 != 0) {
```

Comment:
This check creates a “bifurcation point” in the execution trace - it will follow different branches for different values of actual_values_count.

* Do I understand right it will lead to different assignment tables produced for different inputs?
* Will it cause any downstream implications? E.g. if assignment and transpiling happens with one actual_validators_count, and proof is generated with the other?